publication-rest/src/main/java/no/unit/nva/publication/update/UpdatePublicationHandler.java

```diff
@@ -130,25 +149,149 @@ private static boolean isPending(TicketEntry publishingRequest) {
     }
 
     @Override
-    protected PublicationResponseElevatedUser processInput(UpdatePublicationRequest input,
+    protected PublicationResponseElevatedUser processInput(PublicationRequest input,
                                                            RequestInfo requestInfo,
                                                            Context context)
         throws ApiGatewayException {
+        SortableIdentifier identifierInPath = RequestUtil.getIdentifier(requestInfo);
 
-        var identifierInPath = RequestUtil.getIdentifier(requestInfo);
+        Publication existingPublication = fetchPublication(identifierInPath);
+
+        var userInstance = RequestUtil.createUserInstanceFromRequest(requestInfo, identityServiceClient);
+        var permissionStrategy = PublicationPermissionStrategy.create(existingPublication, userInstance);
+
+        Publication updatedPublication = switch (input) {
+            case UpdatePublicationRequest publicationMetadata ->
+                updateMetadata(publicationMetadata, identifierInPath, existingPublication, permissionStrategy);
+
+            case UnpublishPublicationRequest unpublishPublicationRequest ->
+                unpublishPublication(unpublishPublicationRequest, existingPublication, permissionStrategy);
+
+            case DeletePublicationRequest ignored -> deletePublication(existingPublication, permissionStrategy);
+
+            default -> throw new BadRequestException("Unknown input body type");
+        };
+
+        return PublicationResponseElevatedUser.fromPublication(updatedPublication);
+    }
+
+    private static void throwUnauthorizedUnless(Function<PublicationAction, Boolean> a, PublicationAction requestedPermission) throws UnauthorizedException {
+        if (!a.apply(requestedPermission)) {
+            throw new UnauthorizedException();
+        }
+    }
+
+    private Publication deletePublication(Publication existingPublication,
+                                          PublicationPermissionStrategy permissionStrategy)
+        throws UnauthorizedException, BadRequestException, NotFoundException {
+        throwUnauthorizedUnless(permissionStrategy::hasPermission, PublicationAction.DELETE);
+
+        deleteFiles(existingPublication);
+        resourceService.deletePublication(existingPublication);
+
+        return resourceService.getPublication(existingPublication);
+    }
+
+    private void deleteFiles(Publication publication) {
+        publication.getAssociatedArtifacts()
+            .stream()
+            .filter(File.class::isInstance)
+            .map(File.class::cast)
+            .forEach(
+                file -> s3Driver.deleteFile(UnixPath.of(file.getIdentifier().toString()))
```

Comment:
Perhaps `.map(file::getIdentifier).map(Whatever::toString).map(UnixPath::of)`

Reply:
I like it! 👌